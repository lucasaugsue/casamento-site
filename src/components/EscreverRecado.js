import { Button, Input, Textarea } from '@mantine/core';
import { showNotification } from '@mantine/notifications';
import Email from '@mui/icons-material/Email';
import { InputAdornment, TextField } from '@mui/material';
import { IconAt } from '@tabler/icons';
import React from 'react';
import styles from './EscreverRecado.module.css';


export default function EscreverRecado(){
    const [data, setData] = React.useState({
        email: "", body: ""
    });

    const handleChange = (e) => {
        setData(st => ({
            ...st, [e.target.name]: e.target.value
        }))
    }

    const sendMessage = () => {
        try{
            if(data.email.length < 10) throw new Error("É necessário colocar o seu email!")
            if(data.body.length < 10) throw new Error("É necessário escrever alguma coisa no recado!")
            throw new Error("Funcionalidade ainda não implementada!")

        }catch(err){
            showNotification({message: err.message, color: 'red', autoClose: true})
        }
    }

    return <section className={styles.container} id="recado">
        <div className={styles.whiteBox}>
            <div className={styles.containerText}>
                <div className={styles.gridRecado}>
                    <div className={styles.leftIcon}><Email className={styles.icons}/></div>
                    <div className={styles.textTitle}>Escreva um recado!</div>
                    <div className={styles.rightIcon}> <Email className={styles.icons}/></div>
                </div>
                <div className={styles.subTitle}>Esse espaço é reservado para você que deseja escrever um recado para os noivos.</div>
                <div className={styles.text}>
                    Todo texto será enviado para o nosso e-mail, apenas coloque seu e-mail e o recado que deseja passar para nós.
                </div>
            </div>
            <div className={styles.containerInput}>
                <TextField 
                    name="email"
                    label="Email" 
                    variant="outlined" 
                    placeholder="Seu email"
                    className={styles.input}
                    value={{...data}.email || ""}
                    onChange={(e) => handleChange(e)}
                    // InputProps={{
                    //     startAdornment: (
                    //       <InputAdornment position="start">
                    //         <IconAt />
                    //       </InputAdornment>
                    //     ),
                    // }}
                />

                <TextField 
                    multiline
                    minRows={6}
                    maxRows={6}
                    name="body"
                    label="Escreva o recado aqui" 
                    variant="outlined" 
                    placeholder="Escreva o recado"
                    className={styles.input}
                    value={{...data}.body || ""}
                    onChange={(e) => handleChange(e)}
                />

                <Button 
                    radius="lg"
                    variant="gradient"
                    className={styles.button} 
                    onClick={() => sendMessage()}
                    gradient={{ from: '#f16352', to: '#ec8c69', deg: 35 }}
                >
                    Enviar
                </Button>
            </div>
        </div>
    </section> 
}
