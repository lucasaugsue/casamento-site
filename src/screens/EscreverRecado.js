import { Button, Input, Textarea, Notification } from '@mantine/core';
import { showNotification } from '@mantine/notifications';
import Email from '@mui/icons-material/Email';
import { IconAt, IconUser } from '@tabler/icons';
import React from 'react';
import styles from './EscreverRecado.module.css';
import ClientContext from '../contexts/ClientContext';


export default function EscreverRecado(){
    const { apiRequest } = React.useContext(ClientContext);

    const [loading, setLoading] = React.useState(false);
    const [data, setData] = React.useState({
        nome: "", email: "", 
        recado: ""
    });

    const [sucesso, setSucesso] = React.useState("");
    const [error, setError] = React.useState({
        nome: false, email: false, 
        recado: false, msg: ""
    });


    const handleChange = (e) => {
        setData(st => ({
            ...st, [e.target.name]: e.target.value
        }))
    }

    const validationRecado = () => {
        setError({
            nome: false, email: false, 
            recado: false, msg: ""
        })

        if(data.recado.length < 5) setError((err) => ({...err, recado: true, msg: "É necessário escrever alguma coisa no recado!"}) )
        if(data.email.length < 2) setError((err) => ({...err, email: true, msg: "É necessário colocar o seu email!"}) )
        if(data.nome.length < 2) setError((err) => ({...err, nome: true, msg: "É necessário colocar o seu nome!"}) )
    }

    const sendMessage = () => {
        try{
            validationRecado()

            if( data.nome.length < 2 ||
                data.email.length < 2 ||
                data.recado.length < 5
                ) {  throw new Error(error.msg)
            
            } else {

                setLoading(true)
                apiRequest("POST", "/recados/create", {...data})
                .then((res) => {
                    setLoading(false)
                    setData({nome: "", email: "", recado: ""})
                    setSucesso("Ocorreu tudo certo ao enviar o recado!")
                    showNotification({message: "Recado enviado com sucesso!", color: 'green', autoClose: true})
                })
                .catch((err) => {
                    setLoading(false)
                    showNotification({message: err.message, color: 'red', autoClose: true})
                });
            }

        }catch(err){
            showNotification({message: "Não foi possível realizar a ação!", color: 'red', autoClose: true})
        }
    }

    return <section className={styles.container} id="recado">
        <div className={styles.whiteBox}>
            <div className={styles.containerText}>
                <div className={styles.gridRecado}>
                    <div className={styles.leftIcon}><Email className={styles.icons}/></div>
                    <div className={styles.textTitle}>Escreva um recado!</div>
                    <div className={styles.rightIcon}> <Email className={styles.icons}/></div>
                </div>
                <div className={styles.subTitle}>Esse espaço é reservado para você que deseja escrever um recado para os noivos.</div>
                <div className={styles.text}>
                    Todo texto será enviado para o nosso e-mail, apenas coloque seu e-mail e o recado que deseja passar para nós.
                </div>
            </div>

            {(sucesso.length > 2) && <Notification 
                color="green" 
                radius="md" 
                title="Sucesso!"
                className={styles.notificationCard}
                onClose={() => setSucesso("")}
            > {sucesso} </Notification>}

            {(error.msg.length > 2) && <Notification 
                color="red" 
                radius="md" 
                title="Erro ao enviar o recado!"
                className={styles.notificationCard}
                onClose={() => setError({...error, msg: ""})}
            > {error.msg} </Notification>}

            <div className={styles.containerInput}>
                <Input
                    size="md"
                    radius="lg"
                    name="nome"
                    icon={<IconUser />}
                    placeholder="Seu nome"
                    className={
                        error.nome
                        ? styles.inputError
                        : styles.input
                    }
                    value={{...data}.nome || ""}
                    onChange={(e) => handleChange(e)}
                />

                <Input
                    size="md"
                    radius="lg"
                    name="email"
                    icon={<IconAt />}
                    placeholder="Seu email"
                    className={
                        error.email
                        ? styles.inputError
                        : styles.input
                    }
                    value={{...data}.email || ""}
                    onChange={(e) => handleChange(e)}
                />

                <Textarea
                    size="md"
                    radius="lg"
                    autosize
                    minRows={6}
                    maxRows={6}
                    name="recado"
                    placeholder="Escreva o recado"
                    className={
                        error.recado
                        ? styles.inputError
                        : styles.input
                    }
                    value={{...data}.recado || ""}
                    onChange={(e) => handleChange(e)}
                />

                <Button 
                    radius="lg"
                    variant="gradient"
                    loading={loading}
                    className={styles.button} 
                    onClick={() => sendMessage()}
                    gradient={{ from: '#f16352', to: '#ec8c69', deg: 35 }}
                >
                    Enviar
                </Button>
            </div>
        </div>
    </section> 
}
