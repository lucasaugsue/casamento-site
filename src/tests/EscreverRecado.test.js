import '@testing-library/jest-dom';
import renderer, { act } from 'react-test-renderer';
import ClientContext from '../contexts/ClientContext';
import EscreverRecado from '../screens/EscreverRecado';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';

/**
 * @jest-environment jsdom
 */

const mockApiRequest = jest.fn(() =>
    Promise.resolve({
        item: {
            id: 3,
            nome: 'lucas augsue',
            email: 'lucasaugsue7@gmail.com',
            recado: 'lorem ipslum texto hehe',
            created: '2024-01-02T18:50:52.642Z',
            updated: '2024-01-02T18:50:52.642Z',
        },
        message: 'Criado com sucesso!',
    })
);

describe('<EscreverRecado/>', () => {
	describe('Renderização', () => {
		test('deve renderizar EscreverRecado corretamente', async () => {
			let tree;
		
			await act(async () => {
				const clientContext = { apiRequest: mockApiRequest };

				tree = renderer.create(
					<ClientContext.Provider value={clientContext}>
						<EscreverRecado />
					</ClientContext.Provider>
				);
			});
		
			expect(tree.toJSON()).toMatchSnapshot();
		});
	});

	describe('Interações', () => {
		const clientContext = { apiRequest: mockApiRequest };

		beforeEach(() => {
			render(<ClientContext.Provider value={clientContext}>
				<EscreverRecado />
			</ClientContext.Provider>);
		});

		it('deve usar a funcao sendMessage', async () => {
			const nomeInput = screen.getByPlaceholderText('Seu nome');
			const emailInput = screen.getByPlaceholderText('Seu email');
			const recadoInput = screen.getByPlaceholderText('Escreva o recado');

			fireEvent.change(nomeInput, { target: { value: 'Lucas' } });
			fireEvent.change(emailInput, { target: { value: 'test@example.com' } });
			fireEvent.change(recadoInput, { target: { value: 'Teste de recado' } });

			expect(nomeInput).toHaveValue('Lucas');
			expect(emailInput).toHaveValue('test@example.com');
			expect(recadoInput).toHaveValue('Teste de recado');

			await act(async () => {
				fireEvent.click(screen.getByText('Enviar'));
			});

			expect(nomeInput).toHaveValue('');
		});
	});
});

