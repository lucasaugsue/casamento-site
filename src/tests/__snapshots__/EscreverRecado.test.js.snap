// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EscreverRecado/> Renderização deve renderizar EscreverRecado corretamente 1`] = `
<section
  className="container"
  id="recado"
>
  <div
    className="whiteBox"
  >
    <div
      className="containerText"
    >
      <div
        className="gridRecado"
      >
        <div
          className="leftIcon"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium icons css-i4bv87-MuiSvgIcon-root"
            data-testid="EmailIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M20 4H4c-1.1 0-1.99.9-1.99 2L2 18c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V6c0-1.1-.9-2-2-2zm0 4-8 5-8-5V6l8 5 8-5v2z"
            />
          </svg>
        </div>
        <div
          className="textTitle"
        >
          Escreva um recado!
        </div>
        <div
          className="rightIcon"
        >
           
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium icons css-i4bv87-MuiSvgIcon-root"
            data-testid="EmailIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M20 4H4c-1.1 0-1.99.9-1.99 2L2 18c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V6c0-1.1-.9-2-2-2zm0 4-8 5-8-5V6l8 5 8-5v2z"
            />
          </svg>
        </div>
      </div>
      <div
        className="subTitle"
      >
        Esse espaço é reservado para você que deseja escrever um recado para os noivos.
      </div>
      <div
        className="text"
      >
        Todo texto será enviado para o nosso e-mail, apenas coloque seu e-mail e o recado que deseja passar para nós.
      </div>
    </div>
    <div
      className="containerInput"
    >
      <div
        className="mantine-Input-wrapper input mantine-12sbrde"
      >
        <div
          className="mantine-1hseney mantine-Input-icon"
        >
          <svg
            className="icon icon-tabler icon-tabler-user"
            fill="none"
            height={24}
            stroke="currentColor"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth={2}
            viewBox="0 0 24 24"
            width={24}
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
              stroke="none"
            />
            <circle
              cx={12}
              cy={7}
              r={4}
            />
            <path
              d="M6 21v-2a4 4 0 0 1 4 -4h4a4 4 0 0 1 4 4v2"
            />
          </svg>
        </div>
        <input
          className="mantine-Input-input mantine-Input-withIcon mantine-1h9hasi"
          name="nome"
          onChange={[Function]}
          placeholder="Seu nome"
          value=""
        />
      </div>
      <div
        className="mantine-Input-wrapper input mantine-12sbrde"
      >
        <div
          className="mantine-1hseney mantine-Input-icon"
        >
          <svg
            className="icon icon-tabler icon-tabler-at"
            fill="none"
            height={24}
            stroke="currentColor"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth={2}
            viewBox="0 0 24 24"
            width={24}
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
              stroke="none"
            />
            <circle
              cx={12}
              cy={12}
              r={4}
            />
            <path
              d="M16 12v1.5a2.5 2.5 0 0 0 5 0v-1.5a9 9 0 1 0 -5.5 8.28"
            />
          </svg>
        </div>
        <input
          className="mantine-Input-input mantine-Input-withIcon mantine-1h9hasi"
          name="email"
          onChange={[Function]}
          placeholder="Seu email"
          value=""
        />
      </div>
      <div
        className="mantine-InputWrapper-root mantine-Textarea-root input mantine-1m3pqry"
      >
        <div
          className="mantine-Input-wrapper mantine-Textarea-wrapper mantine-12sbrde"
        >
          <textarea
            aria-invalid={false}
            className="mantine-Input-input mantine-Textarea-input mantine-1jc3cvc"
            id="mantine-r0"
            name="recado"
            onChange={[Function]}
            placeholder="Escreva o recado"
            rows={6}
            value=""
          />
        </div>
      </div>
      <button
        className="mantine-UnstyledButton-root mantine-Button-root button mantine-14dldul"
        data-button={true}
        onClick={[Function]}
        type="button"
      >
        <div
          className="mantine-3xbgk5 mantine-Button-inner"
        >
          <span
            className="mantine-qo1k2 mantine-Button-label"
            style={
              {
                "textTransform": undefined,
              }
            }
          >
            Enviar
          </span>
        </div>
      </button>
    </div>
  </div>
</section>
`;
